src/rpc/server/linux_io_uring.zig

```diff
@@ -0,0 +1,835 @@
+    misc_connection_failure,
+
+    /// Failed to parse the HTTP request head, perhaps because it was too big,
+    /// the request target was too long, or because it was invalid in some way.
+    http_head_parse_error,
+
+    /// Somehow the splice arguments or their associated state were invalid.
+    splice_error,
+};
+
+const ConsumeOurCqeError =
+    IouGetSqeError ||
+    HandleRecvBodyError ||
+    std.mem.Allocator.Error ||
+    connection.HandleAcceptError ||
+    connection.HandleRecvError ||
+    connection.HandleSendError ||
+    connection.HandleSpliceError;
+
+/// Panic message for handling `EAGAIN`; we're not using nonblocking sockets at all,
+/// so it should be impossible to receive that error, or for such an error to be
+/// triggered just from malicious connections.
+const EAGAIN_PANIC_MSG =
+    "The file/socket should not be in nonblocking mode;" ++
+    " server or file/socket configuration error.";
+
+/// On return, `cqe.user_data` is in an undefined state - this is to say,
+/// it has either already been `deinit`ed, or it has been been re-submitted
+/// in a new `SQE` and should not be modified; in either scenario, the caller
+/// should not interact with it.
+fn consumeOurCqe(
+    liou: *LinuxIoUring,
+    server_ctx: *server.Context,
+    cqe: OurCqe,
+) ConsumeOurCqeError!ConsumeOurCqeResult {
+    const entry = cqe.user_data;
+    errdefer entry.deinit(server_ctx.allocator);
+
+    const entry_data: *EntryData = entry.ptr orelse {
+        // `accept_multishot` cqe
+
+        // we may need to re-submit the `accept_multishot` sqe.
+        const accept_cancelled = cqe.flags & std.os.linux.IORING_CQE_F_MORE == 0;
+        if (accept_cancelled) liou.multishot_accept_submitted = false;
+
+        switch (try connection.handleAcceptResult(cqe.err())) {
+            .success => {},
+            // just quickly exit; if we need to re-issue, that's already handled above
+            .again => return .ok,
+            .intr => return .signal_interrupted_operation,
+            .conn_aborted => return .misc_connection_failure,
+        }
+
+        const stream: std.net.Stream = .{ .handle = cqe.res };
+        errdefer stream.close();
+
+        server_ctx.wait_group.start();
+        errdefer server_ctx.wait_group.finish();
+
+        const buffer = try server_ctx.allocator.alloc(u8, server_ctx.read_buffer_size);
+        errdefer server_ctx.allocator.free(buffer);
+
+        const new_recv_entry: Entry = entry: {
+            const data_ptr = try server_ctx.allocator.create(EntryData);
+            errdefer comptime unreachable;
+
+            data_ptr.* = .{
+                .buffer = buffer,
+                .stream = stream,
+                .state = EntryState.INIT,
+            };
+            break :entry .{ .ptr = data_ptr };
+        };
+        errdefer if (new_recv_entry.ptr) |data_ptr| server_ctx.allocator.destroy(data_ptr);
+
+        _ = liou.io_uring.recv(
```

Comment:
Just thinking out loud here... Imagine if io_uring had an sqe called "accept_multishot_recv" that didn't bother with a cqe for the accept, and the kernel would just assume automatically that we want to receive the data, and userspace wouldn't be bothered until after a recv cqe was ready. That way we could avoid these steps here. Do you think something like that would be useful?

Reply:
That seems like it would be useful, yeah, but I don't really think there's a way to represent that; can't link the SQEs because the `recv` SQE needs to know immediately what socket it's receiving from, and there's no mechanism to represent using a handle that's only later created or anything like that.